File: python/qlambda/license_policy.py
"""Array-backed tri-license policy engine.

This keeps repository language metrics aligned with the actual Python/Lean
package.
"""

from __future__ import annotations

import argparse
from typing import Dict, Tuple

LICENSES: Tuple[str, ...] = ("bsl_1_1", "agpl_3_0", "mpl_2_0", "commercial")

USE_CASES: Dict[str, str] = {
    "saas_wrapper": "agpl_3_0",
    "enterprise_restricted": "bsl_1_1",
    "file_level_mod": "mpl_2_0",
    "copyleft_bypass": "commercial",
    "open_source_redistribution": "agpl_3_0",
}

COMPATIBILITY: Tuple[Tuple[str, str], ...] = (
    ("mpl_2_0", "proprietary"),
    ("mpl_2_0", "mpl_2_0"),
    ("bsl_1_1", "source_available"),
    ("agpl_3_0", "agpl_3_0"),
    ("commercial", "proprietary"),
)


def select_license(use_case: str) -> str:
    try:
        return USE_CASES[use_case]
    except KeyError as exc:
        raise ValueError(f"unknown use case: {use_case}") from exc


def is_compatible(license_name: str, dependency_type: str) -> bool:
    return (license_name, dependency_type) in COMPATIBILITY


def main() -> None:
    parser = argparse.ArgumentParser(description="Select or check the tri-license policy.")
    sub = parser.add_subparsers(dest="cmd", required=True)

    select = sub.add_parser("select")
    select.add_argument("use_case")

    check = sub.add_parser("check")
    check.add_argument("license")
    check.add_argument("dependency")

    sub.add_parser("matrix")
    args = parser.parse_args()

    if args.cmd == "select":
        print(f"Recommended License: {select_license(args.use_case)}")
    elif args.cmd == "check":
        ok = is_compatible(args.license, args.dependency)
        label = "compatible" if ok else "INCOMPATIBLE"
        print(f"{args.license} is {label} with {args.dependency}.")
        raise SystemExit(0 if ok else 1)
    elif args.cmd == "matrix":
        for license_name, dependency in COMPATIBILITY:
            print(f"{license_name} <-> {dependency}")


if __name__ == "__main__":
    main()

File: backends/license_policy.pl
% =====================================================================
% LICENSE POLICY ENGINE (Prolog Backend)
% =====================================================================

% Define available licenses
license(bsl_1_1).
license(agpl_3_0).
license(mpl_2_0).
license(commercial).

% Define use cases and map them to the optimal license tier
use_case(saas_wrapper, agpl_3_0).
use_case(enterprise_restricted, bsl_1_1).
use_case(file_level_mod, mpl_2_0).
use_case(copyleft_bypass, commercial).
use_case(open_source_redistribution, agpl_3_0).

% Compatibility matrix: compatible(LicenseA, LicenseB)
compatible(mpl_2_0, proprietary).
compatible(mpl_2_0, mpl_2_0).
compatible(bsl_1_1, source_available).
compatible(agpl_3_0, agpl_3_0).
compatible(commercial, proprietary).

% Select license based on use case query
select_license(UseCase, SelectedLicense) :-
    use_case(UseCase, SelectedLicense).

% Validate dependency compatibility
check_compatibility(License, DependencyType) :-
    compatible(License, DependencyType),
    format('~w is compatible with ~w.~n', [License, DependencyType]).

check_compatibility(License, DependencyType) :-
    \+ compatible(License, DependencyType),
    format('WARNING: ~w is INCOMPATIBLE with ~w.~n', [License, DependencyType]),
    fail.

% CLI Entrypoint handlers
main :-
    current_prolog_flag(argv, Argv),
    handle_args(Argv).

handle_args(['matrix']) :-
    write('=== LICENSE COMPATIBILITY MATRIX ===\n'),
    forall(compatible(A, B), format(' [OK] ~w <--> ~w\n', [A, B])),
    halt.

handle_args(['select', UseCase]) :-
    atom_string(UseCaseAtom, UseCase),
    ( select_license(UseCaseAtom, License)
    -> format('Recommended License: ~w\n', [License])
    ; format('Unknown use case: ~w\n', [UseCase])
    ),
    halt.

handle_args(['check', License, Dep]) :-
    atom_string(LicAtom, License),
    atom_string(DepAtom, Dep),
    check_compatibility(LicAtom, DepAtom),
    halt.

handle_args(_) :-
    write('Usage: swipl -q -t halt -f license_policy.pl -- [matrix | select <use_case> | check <license> <dep>]\n'),
    halt.

:- initialization(main, main).
